modules/ding_campaign/ding_campaign.bpi.inc

```diff
@@ -0,0 +1,108 @@
+<?php
+
+/**
+ * @file
+ * Implementation of BPI hooks.
+ */
+
+/**
+ * Implements hook_bpi_syndicate_get_bpi_types().
+ */
+function ding_campaign_bpi_syndicate_get_bpi_types() {
+  return array('campaign');
+}
+
+/**
+ * Implements bpi_syndicate_action_url().
+ */
+function ding_campaign_bpi_syndicate_action_url($bpi_type, $bpi_id, array $mapping) {
+  if (in_array($bpi_type, ding_campaign_bpi_syndicate_get_bpi_types())) {
+    return url(str_replace('_', '-', 'node/add/' . $mapping['bpi_content_type']), array('query' => array('bpi_type' => $bpi_type, 'bpi_id' => $bpi_id)));
+  }
+
+  return NULL;
+}
+
+/**
+ * Implements hook_bpi_get_field_mapping().
+ */
+function ding_campaign_bpi_get_field_mapping_form($bpi_type, $content_type, array $field_mapping) {
+  if (in_array($bpi_type, ding_campaign_bpi_syndicate_get_bpi_types())) {
+    $field_options = array('' => '') + bpi_get_field_instances($content_type);
+
+    return array(
+      'bpi_field_title' => array(
+        '#type' => 'select',
+        '#title' => t('BPI title'),
+        '#description' => t('Titles are automatically assigned.'),
+        '#options' => array('title' => t('Title')),
+        '#value' => 'title',
+        '#disabled' => TRUE,
+      ),
+
+      'bpi_field_body' => array(
+        '#type' => 'select',
+        '#title' => t('BPI body'),
+        '#description' => t('Field to extract the main content from (body field).'),
+        '#options' => $field_options,
```

Comment:
When looking at this, `bpi_bpi_get_field_mapping_form()` and the way the form is handled it seems as if a quite specific structure is needed for this to work. This does not go very well with a hook that basicly allows you to provide whatever form API structure you want.

I suggest replacing `bpi_get_field_mapping_form` with a hook which returns an array of BPI field data (names, descriptions + ?) and then that is used to generate the form API.

Example:

```php
function ding_campaign_bpi_get_fields() {
   return array(
     bpi_field_body => array(
       'title' => t('BPI body'),
       'description' => t('Field to extract the main content from (body field).'),
     ),
     // and so on...
   )
}